src/ranged.cpp

```diff
@@ -2695,10 +2697,78 @@ void target_ui::apply_aim_turning_penalty()
 
 void target_ui::action_switch_mode()
 {
-    relevant->gun_cycle_mode();
-    if( relevant->gun_current_mode().flags.count( "REACH_ATTACK" ) ) {
-        relevant->gun_cycle_mode();
+    uilist menu;
+    menu.settext( _( "Select preferences" ) );
+    const std::pair<int, int> aim_modes_range = std::make_pair( 0, 100 );
+    const std::pair<int, int> firing_modes_range = std::make_pair( 100, 200 );
+
+
+    // list aiming modes
+    menu.addentry( -1, false, 0, "  " + std::string( _( "Default aiming mode" ) ) );
+    menu.entries.back().force_color = true;
+    menu.entries.back().text_color = c_cyan;
+
+    for( auto it = aim_types.begin(); it != aim_types.end(); ++it ) {
+        const bool is_active_aim_mode = aim_mode == it;
+        const std::string text = ( it->name.empty() ? _( "Immediate" ) : it->name ) +
+                                 ( is_active_aim_mode ? _( " (active)" ) : "" );
+        menu.addentry( aim_modes_range.first + std::distance( aim_types.begin(), it ),
+                       true, MENU_AUTOASSIGN, text );
+        if( is_active_aim_mode ) {
+            menu.entries.back().text_color = c_light_green;
+        }
+    }
+
+    menu.addentry( -1, false, 0, "  " + std::string( _( "Firing mode" ) ) );
+    menu.entries.back().force_color = true;
+    menu.entries.back().text_color = c_cyan;
+    if( relevant->is_gun() || relevant->is_gunmod() ) {
+        const std::map<gun_mode_id, gun_mode> gun_modes = relevant->gun_all_modes();
```

Comment:
This check is always `true`, since the code that launches target ui already checks whether the item is a weapon.
Also, iirc `gun_all_modes` does the same checks anyway.

Reply:
Thanks I stole that from somewhere without checking